res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+  var data = [];
+  for (var i = 0; i < 38; i++) {
+    data[i] = 0;
+  }
+  // Leave USB plug indicator light on.
+  data[0x1C] = 0x7f;
+  controller.send(data, data.length, 0x80);
+  
+  for (i = 0; i < 33; i++) {
+    data[i] = 0;
+  }
+  controller.send(data, data.length, 0x81);
+}
+
+TraktorS2MK2.incomingData = function(data, length) {
+  TraktorS2MK2.controller.parsePacket(data, length);
+}
+
+// The short message handles buttons and jog wheels.
+TraktorS2MK2.shortMessageCallback = function(packet, data) {
+  for (name in data) {
+    field = data[name];
+    if (field.name === "!jog_wheel") {
+      TraktorS2MK2.controller.processControl(field);
+      continue;
+    }
+
+    TraktorS2MK2.controller.processButton(field);
+  }
+}
+
+// There are no buttons handled by the long message, so this is a little simpler.  (Even though
+// this is very similar to the other handler, it's easier to keep them separate to know what's
+// a control and what's a button.
+TraktorS2MK2.longMessageCallback = function(packet, data) {
+  for (name in data) {
+    field = data[name];
+    TraktorS2MK2.controller.processControl(field);
+  }
+}
+
+TraktorS2MK2.samplerGainKnob = function(field) {
+  for (var i = 1; i <= 8; i++) {
+    engine.setParameter("[Sampler" + i + "]", "pregain", field.value / 4096);
+  }
+}
+
+TraktorS2MK2.toggleButton = function(field) {
+  if (field.value > 0) {
+    script.toggleControl(field.group, field.name);
+  }
+}
+
+TraktorS2MK2.shift = function(field) {
+  var group = field.id.split(".")[0];
+  TraktorS2MK2.shift_pressed[group] = field.value;
+  TraktorS2MK2.outputCallback(field.value, field.group, "!shift");
+}
+
+TraktorS2MK2.loadTrack = function(field) {
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  if (TraktorS2MK2.shift_pressed[group]) {
+    engine.setValue(field.group, "eject", field.value);
+  } else {
+    engine.setValue(field.group, "LoadSelectedTrack", field.value);
+  }
+}
+
+TraktorS2MK2.syncEnabled = function(field) {
+  var now = Date.now();
+
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  // If shifted, just toggle.
+  // TODO(later version): actually make this enable explicit master.
+  if (TraktorS2MK2.shift_pressed[group]) {
+    if (field.value === 0) {
+      return;
+    }
+    var synced = engine.getValue(field.group, "sync_enabled");
+    engine.setValue(field.group, "sync_enabled", !synced);
+  } else {
+    if (field.value === 1) {
+      TraktorS2MK2.sync_enabled_time[field.group] = now;
+      engine.setValue(field.group, "sync_enabled", 1);
+    } else {
+      var cur_enabled = engine.getValue(field.group, "sync_enabled");
+      if (!cur_enabled) {
+        // If disabled, and switching to disable... stay disabled.
+        engine.setValue(field.group, "sync_enabled", 0);
+        return;
+      }
+      // was enabled, and button has been let go.  maybe latch it.
+      if (now - TraktorS2MK2.sync_enabled_time[field.group] > 300) {
```

Comment:
this shouldn't be necessary I think. Just bind value > 0 to sync_enabled and the latching should work without any hacks.

Reply:
That's not how it works, but I think that's how it should work...

Reply:
as I pointed out on other pull requests, controller scripts shouldn't change mixxx's behavior due to personal preferences of script developers.  We should keep the bindings consistent.  Then, when we change how the sync button works (if we want to remove latching) then all the controller scripts will get the update.  For now, just bind to sync_enabled and don't use your own latching code, or else we'll be stuck with it! (we will have to decide how to bind momentary sync, auto sync, and explicit master for 2.4)

Reply:
> controller scripts shouldn't change mixxx's behavior due to personal preferences of script developers

I agree, which is why I'm leaving it as is.

>  For now, just bind to sync_enabled and don't use your own latching code, or else we'll be stuck with it! (we will have to decide how to bind momentary sync, auto sync, and explicit master for 2.4)

As I said above, that doesn't work as you think it does and it never has.

Reply:
on the vci-400, the sync buttons are bound like this:

```
<control>
   <group>[Channel1]</group>
   <key>sync_enabled</key>
   <description>MIDI Learned from 22 messages.</description>
   <status>0x92</status>
   <midino>0x01</midino>
   <options>
      <normal/>
   </options>
</control>
```

Reply:
This is an HID controller.

Reply:
Ah, the problem is that the latching effect was implemented as part of the midi button handler code, not the controlobject code.  So that's why it doesn't work over HID by setting the control object explicitly.